packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -407,17 +423,30 @@ export class ConnectionManager implements IConnectionManager {
         this.connectCore(connectionMode).catch((error) => {
             const normalizedError = normalizeError(error, { props: fatalConnectErrorProp });
             this.props.closeHandler(normalizedError);
+            this.pendingConnection = undefined;
         });
     }
 
     private async connectCore(connectionMode?: ConnectionMode): Promise<void> {
         assert(!this.closed, 0x26a /* "not closed" */);
 
-        if (this.connection !== undefined || this.pendingConnection) {
-            return;
+        if (this.connection !== undefined) {
+            return;  // Connection attempt already completed successfully
+        }
+
+        let requestedMode = connectionMode;
+        if (this.pendingConnection !== undefined) {
+            // If there is no specified ConnectionMode, then default to the previous mode
+            if (requestedMode === undefined) {
+                requestedMode = this.pendingConnection.connectionMode;
+            }
+            this.cancelConnection();  // Throw out in-progress connection attempt in favor of new attempt
         }
 
-        let requestedMode = connectionMode ?? this.defaultReconnectionMode;
+        requestedMode = requestedMode ?? this.defaultReconnectionMode;
+
+        this.pendingConnection = { abortController: new AbortController(), connectionMode: requestedMode };
```

Comment:
Hmm.  The downside is that pendingConnection doesn't reflect the actual mode used in the pending connection if requestedMode is changed below.

I would take a closer look at what tests failed and why, and see if there's a way to resolve this that's more cohesive.

TBH I'm not sure I'm following your point about "so we can keep it if that attempt gets cancelled", so maybe this is fine.

Reply:
Another thing to consider is something I observed previously in this code, which is that I believe as-is we only ever "upgrade" a read to a write connection.  You have to trace around kind of a lot of places to confirm this.  I wonder if that semantic should be enforced better.  Right now there are at least 2-3 different fallbacks, and you're adding another, and it's kind of complex.  If it were all about `||`ing together everyone's opinion of whether write was required, it would be a little more robust.

Reply:
I looked into it more and found why the tests are failing. The root cause and why we probably shouldn't continuously modify `this.pendingConnection.connectionMode` is similar to the issue [here](https://github.com/microsoft/FluidFramework/pull/10493/files#r906296784). Essentially, if we call `connect()` twice in a row we will abort the first connection attempt. However, at certain points in the first attempt's `connectCore()` we may not have reached the check to see if it was aborted and may try to modify `this.pendingConnection`, which may be undefined (due to being aborted) or even belong to the second (non-aborted) connection attempt. I think it makes sense to keep passing down the local `requestedMode` variable to avoid issues, similar to how we reference the local `abortSignal` variable.

Reply:
Ok - got it!  This is subtle!  Only took me about 20 minutes reading and re-reading the code after hearing your explanation haha.  The point is, as soon as `connectCore` goes async `this.pendingConnection` can't be "trusted" anymore, since `connect()` is fire-and-forget.  I was trying to think if there's a way to structure the code to encapsulate this better but not seeing anything. Something like pulling the loop from connectCore into a static function to highlight the fact that it probably shouldn't be modifying state since there can be concurrent loops.  Might work but not necessary for this PR to avoid over-complicating the change.

Here are some changes I would suggest though:

1. Add a comment on `pendingConnection` (or `connectCore` or both?) that explains something about how when the connection code goes async, `pendingConnection` may have changed by the time the async code continues.
2. Change `IPendingConnection` to have an `abort()` function rather than holding the whole `AbortController`. This way there's no temptation to try to check its state elsewhere in the code.
3. Consider if `IPendingConnection.connectionMode` is actually necessary.  I'm a little doubtful at this point, and dropping it would avoid some confusion IMO.
4. I would flip the order of things so the last thing you do before entering The Loop is initialize `this.pendingConnection` (same place `this.pendingConnection` is set on the LHS)

That's my 2 cents anyway! Thanks for clarifying!